### PR TITLE
Projectile Refactor

### PR DESCRIPTION
**Pull request by OmKurkure**
Changes:
1. Added Equippables for the equippable items like weapons and potions
2. Refactored projectiles. Each projectile now handles its own setup and collidable creation
3. Can place multiple bombs. All other projectiles are single instances only because multiple sword beams or arrows make Link too overpowered
4. Added LinkDeath Animation and WallMasterGrab Animation code (not tested). For the wallMasterGrab Animation, the logic is there, however it does not work because there are some issues with the wallmaster enemy class, so testing of the animation was not possible.
5. Added UI functionality to cycle and equip weapons and potions using the B key
6. New item: Red Potion -> gives full health and a blue potion 
7. Made some changes to eventManager. It is static right now, but can be removed. The alternative to handling player death, triforce pickup, and wallmaster animation is to send in the sceneManager(previously DungeonScene) reference to classes which will increase coupling, but will allow classes to change the activeScene to play the required animations.
8. Added the compass and Map sprite to inventory to indicate the map and compass have been picked up

### CODE REVIEW

File: MonoZelda/Link/Projectiles/ProjectileManager.cs [hunk 1]
```diff
@@ -1,213 +1,93 @@
...
-        if (EquippedProjectile == ProjectileType.CandleBlue && PlayerState.IsCandleUsed)
-        {
-            return; 
+            if (ActiveProjectiles.ContainsKey(projectileType) == false)
+            {
+                IProjectile projectile = projectileFactory.GetProjectileObject(projectileType, collisionController);
+                ActiveProjectiles.Add(projectileType, projectile);
+                DeductPlayerResources(projectileType);
+                projectile.Setup();
+            }
```
Comment (xsedim) [inline MonoZelda/Link/Projectiles/ProjectileManager.cs:67]: Big improvement in abstracting projectile types in order to make `ProjectileManager`'s job easier

File: MonoZelda/UI/HUDMapWidget.cs [hunk 2]
```diff
@@ -42,6 +43,12 @@ public HUDMapWidget(Screen screen, Point position) : base(screen, position)
         compassMarker = new(SpriteType.HUD, SpriteLayer.HUD + 1, WidgetLocation + compassMarkerOffset);
         compassMarker.SetSprite("map_marker_red_blinking");
         compassMarker.Enabled = compassMarkerEnabled;
+        compassItem = new(SpriteType.HUD, SpriteLayer.HUD + 1, WidgetLocation + mapItemSpriteOffset);
+        compassItem.SetSprite("compass");
+        compassItem.Enabled = compassMarkerEnabled;
+        mapItem = new(SpriteType.HUD, SpriteLayer.HUD + 1, WidgetLocation + compassItemSpriteOffset);
+        mapItem.SetSprite("map");
+        mapItem.Enabled = mapEnabled;
```
Comment (xsedim) [inline MonoZelda/UI/HUDMapWidget.cs:51]: The compass and map sprites should go in `InventoryMapWidget`, not `HUDMapWidget`. The HUD is the part of the inventory you see during normal gameplay, and those sprites are only displayed in the full screen inventory.

**Review by xsedim — COMMENTED** [review]
Overall a significant number of fixes, improvements, and refactoring was done in this branch, so great work on all of that. I've only done a code review, not tested the branch for bugs or anything, so that should still done before merging.

One comment not related to code code quality or functionality: this is a very large PR that changed a lot of game systems at once and changed a lot of files, so in the future try to make your PRs composed of smaller, distinct changes. That will make it easier on you to code discrete issues one at a time, easier on the rest of the team to review the PR, and easier on the repo by introducing less significant conflicts with other branches. For example, you listed 8 major changes in the PR, but they probably could have been 8 separate PRs of smaller fixes and changes that would have been easier to handle one at a time.

Also, Josh's hot save/load PR has been merged which I'm guessing is where a lot of the current conflicts are coming from, so some refactoring will need to be done so that the new equippable inventory system is saved and loaded correctly. I think this shouldn't be too hard by using Josh's serialization classes.

File: MonoZelda/Events/EventHandlers.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    public  void TriggerLinkDeathAnimation()
+    {
+        LinkDeath?.Invoke();
+    }
+
+    public void TriggerWallMasterGrabAnimation()
+    {
+        WallmasterGrab?.Invoke();
+    }
+}
```
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Events/EventHandlers.cs:30]: I totally forgot that I could have used the commandManager for this. The eventManager gave me a lot of problems during bug fixing

File: MonoZelda/Items/ItemClasses/RedPotion.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        base.ItemSpawn(itemSpawn, collisionController);
+        itemDict.SetSprite("potion");
+    }
+
+    public override void HandleCollision(CollisionController collisionController)
+    {
+        if (!PlayerState.EquippableInventory.Contains(EquippableType.BluePotion))
+        {
+            PlayerState.EquippableInventory.Add(EquippableType.RedPotion);
+        }
```
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Items/ItemClasses/RedPotion.cs:29]: Thats a good idea

File: MonoZelda/Commands/GameCommands/PlayerUseEquippableCommand.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    public PlayerUseEquippableCommand(EquippableManager equippableManager, PlayerSpriteManager player)
+    {
+        this.player = player;
+        this.equippableManager = equippableManager;
+    }
+
+    public void Execute(params object[] metadata)
+    {
+        Keys pressedKey = (Keys)metadata[0];
```
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Commands/GameCommands/PlayerUseEquippableCommand.cs]: Fixed

File: MonoZelda/Events/EventHandlers.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Events/EventHandlers.cs]: Fixed

File: MonoZelda/Items/ItemClasses/RedPotion.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Items/ItemClasses/RedPotion.cs]: Added this feature

File: MonoZelda/Link/Equippables/EquippableManager.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    public void UseEquippedItem()
+    {
+        if (PlayerState.EquippedItem == EquippableType.None)
+        {
+            return;
+        }
+        else
+        {
+            equippableObjects[PlayerState.EquippedItem].Use(projectileManager,this);
+        }
```
[the remark this answers is not in the document]
Comment (OmKurkure) [reply MonoZelda/Link/Equippables/EquippableManager.cs]: Fixed

**Pull request merged**